### DISCUSSION TITLE
Possible to blacklist packages?

### DISCUSSION THREAD

**Discussion by mathomp4** [Homebrew/discussions#1908] [Everyday usage]
I was wondering if it's possible with brew to "blacklist" packages? The main reason I ask is that I do code development on my Macbook and I like to control my compilers by building them myself and maintaining Lua modules to enable them. 

Thus, I tend to be very careful and never let brew install `gcc` since I want to maintain that myself and make sure it's not in my "default" path (which has the brew bin). I have to live with `clang` being in my default PATH, but `gcc` only appears when I ask it to.

But the fun part is when I do a `brew install foo` and find it wants `gcc` and boom, I now have `gcc-11` or whatever in my path. I try to look at `brew info foo` first, but sometimes I see a cool new package and get impatient. 😄 

So I wondered, is there a way to tell `brew` "never install `gcc` or any package that depends on it"? I don't mind losing out on some packages if it means I can keep my control.

**Comment by SMillerDev** [comment]
You could just install it and `brew unlink gcc`, that way it won't show up in your path.

**Reply by mathomp4** [reply]
Hmm. I hadn't thought of that...

**Reply by pan3793** [reply]
> You could just install it and `brew unlink gcc`, that way it won't show up in your path.

This is a workaround, we still need to install the unused deps and upgrade them when run `brew upgrade`. Is there any concerns to add a blacklist feature?

**Reply by micolous** [reply]
If there was a way to tell Homebrew to `unlink` the package automatically (ie: even after upgrades), then it would allow any formula which depended on it to build with the Homebrew version, while not affecting things outside of Homebrew.

**Comment by pan3793** [comment]
I'm also looking forward to this feature, lots of Java softwares have been declared depend on `openjdk`, but most of them can not work with the latest version of openjdk, e.g. https://github.com/Homebrew/homebrew-core/issues/98294

**Reply by agenslerDNA** [reply]
I also would love to see this feature. I'm running into the exact same type of scenarios. I find myself having to fix my environment more and more, so a way to blacklist would be a great QOL addition.

**Reply by SMillerDev** [reply]
Alternatively, submitting pull requests to ensure the tests for formulae actually catch the version issues in CI make the Homebrew packages better for everyone.

**Comment by iesahin** [comment]
This is also common for Rust especially when certain code requires nightly and I switch release channels use rustup. Brew installs the stable, rustup installs nightly and when a crate requires nightly, Brew's Rust interferes to compile. If you forget to uninstall/unlink Brew's Rust, you get weird errors, especially for beginners. 

There must be a way to tell "I already have this package, please don't install and go on assuming you have" to Brew.

**Reply by SMillerDev** [reply]
There is not, because brew can only know about packages it installed. It sounds like you'd be better off just using rustup to manage your rust install
